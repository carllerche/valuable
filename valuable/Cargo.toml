[package]
name = "valuable"
version = "0.1.0"
authors = ["Carl Lerche <me@carllerche.com>"]
edition = "2018"
license = "MIT"
description = "Values"

[features]
default = ["std"]

# Provide derive(Valuable) macro
derive = ["valuable-derive"]

# Provide impls for common standard library types like Vec<T> and HashMap<K, V>.
std = ["alloc"]

# Provide imps for types in Rust's `alloc` library.
alloc = []

# Provides serde_json::Value instrument for valuable
json = ["serde_json"]

[dependencies]
valuable-derive = { version = "0.1.0", optional = true, path = "../valuable-derive" }
serde_json = { version = "1.0.46", optional = true }

[dev-dependencies]
criterion = "0.3"

[[bench]]
name = "structable"
harness = false
