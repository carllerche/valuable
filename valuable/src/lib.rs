mod enumerable;
pub use enumerable::{Enumerable, EnumDef, VariantDef};

pub mod field;

mod listable;
pub use listable::Listable;

/*
mod mappable;
pub use mappable::Mappable;
*/

mod record;
pub use record::NamedValues;

mod structable;
pub use structable::{Structable, StructDef};

mod valuable;
pub use valuable::Valuable;

mod value;
pub use value::Value;

mod visit;
pub use visit::Visit;

#[cfg(feature = "derive")]
pub use valuable_derive::*;
