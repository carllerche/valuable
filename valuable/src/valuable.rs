use crate::{Slice, Value, Visit};

use core::fmt;

pub trait Valuable {
    fn as_value(&self) -> Value<'_>;

    fn visit(&self, visit: &mut dyn Visit);

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        const N: usize = 8;

        let mut batch: [_; N] = Default::default();
        let mut curr = 0;

        if slice.is_empty() {
            visit.visit_slice(Slice::Value(&batch[..0]));
            return;
        }

        for v in slice {
            if curr == N {
                visit.visit_slice(Slice::Value(&batch[..]));
                curr = 0;
            }

            batch[curr] = v.as_value();
            curr += 1;
        }

        if curr > 0 {
            visit.visit_slice(Slice::Value(&batch[..curr]));
        }
    }
}

impl<V: ?Sized + Valuable> Valuable for &V {
    fn as_value(&self) -> Value<'_> {
        (*self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(*self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::boxed::Box<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::rc::Rc<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::sync::Arc<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

macro_rules! valuable {
    (
        $(
            $variant:ident($ty:ty),
        )*
    ) => {
        $(
            impl Valuable for $ty {
                fn as_value(&self) -> Value<'_> {
                    Value::$variant(*self)
                }

                fn visit(&self, visit: &mut dyn Visit) {
                    visit.visit_value(self.as_value());
                }

                fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
                where
                    Self: Sized,
                {
                    visit.visit_slice(Slice::$variant(slice));
                }
            }

        )*
    };
}

valuable! {
    Bool(bool),
    Char(char),
    F32(f32),
    F64(f64),
    I8(i8),
    I16(i16),
    I32(i32),
    I64(i64),
    I128(i128),
    Isize(isize),
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    Usize(usize),
}

impl Valuable for () {
    fn as_value(&self) -> Value<'_> {
        Value::Unit
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::Unit);
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::Unit(slice));
    }
}

impl Valuable for &'_ str {
    fn as_value(&self) -> Value<'_> {
        Value::String(self)
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::String(self));
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::Str(slice));
    }
}

#[cfg(feature = "alloc")]
impl Valuable for alloc::string::String {
    fn as_value(&self) -> Value<'_> {
        Value::String(&self[..])
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::String(self));
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::String(slice));
    }
}

#[cfg(feature = "std")]
impl Valuable for dyn std::error::Error + '_ {
    fn as_value(&self) -> Value<'_> {
        Value::Error(self)
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(self.as_value());
    }
}

impl fmt::Debug for dyn Valuable + '_ {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = self.as_value();
        value.fmt(fmt)
    }
}
