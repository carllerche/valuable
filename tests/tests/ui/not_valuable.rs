use valuable::Valuable;

struct S;

#[derive(Valuable)]
struct Struct {
    f: Option<S>,
}

#[derive(Valuable)]
struct Tuple(Option<S>);

#[derive(Valuable)]
enum Enum {
    Struct { f: Option<S> },
    Tuple(Option<S>),
}

fn main() {}
