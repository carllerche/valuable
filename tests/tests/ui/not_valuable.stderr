error[E0277]: the trait bound `S: Valuable` is not satisfied
  --> tests/ui/not_valuable.rs:7:8
   |
7  |     f: Option<S>,
   |        ^^^^^^^^^ the trait `Valuable` is not implemented for `S`
   |
   = note: required because of the requirements on the impl of `Valuable` for `Option<S>`
note: required by `as_value`
  --> $WORKSPACE/valuable/src/valuable.rs
   |
   |     fn as_value(&self) -> Value<'_>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `S: Valuable` is not satisfied
  --> tests/ui/not_valuable.rs:11:14
   |
11 | struct Tuple(Option<S>);
   |              ^^^^^^^^^ the trait `Valuable` is not implemented for `S`
   |
   = note: required because of the requirements on the impl of `Valuable` for `Option<S>`
note: required by `as_value`
  --> $WORKSPACE/valuable/src/valuable.rs
   |
   |     fn as_value(&self) -> Value<'_>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `S: Valuable` is not satisfied
  --> tests/ui/not_valuable.rs:15:17
   |
15 |     Struct { f: Option<S> },
   |                 ^^^^^^^^^ the trait `Valuable` is not implemented for `S`
   |
   = note: required because of the requirements on the impl of `Valuable` for `Option<S>`
   = note: 1 redundant requirements hidden
   = note: required because of the requirements on the impl of `Valuable` for `&Option<S>`
note: required by `as_value`
  --> $WORKSPACE/valuable/src/valuable.rs
   |
   |     fn as_value(&self) -> Value<'_>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `S: Valuable` is not satisfied
  --> tests/ui/not_valuable.rs:16:11
   |
16 |     Tuple(Option<S>),
   |           ^^^^^^^^^ the trait `Valuable` is not implemented for `S`
   |
   = note: required because of the requirements on the impl of `Valuable` for `Option<S>`
   = note: 1 redundant requirements hidden
   = note: required because of the requirements on the impl of `Valuable` for `&Option<S>`
note: required by `as_value`
  --> $WORKSPACE/valuable/src/valuable.rs
   |
   |     fn as_value(&self) -> Value<'_>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
